driver.go

```diff
@@ -2,88 +2,180 @@ package main
+			"com.docker.imagefs.target":  r.Options["target"],
+		},
+	}
+	// TODO handle error
+	hostConfig := &container.HostConfig{
+		Binds: []string{"/tmp/runtime:/runtime"},
+	}
+	networkConfig := &network.NetworkingConfig{}
+	cont, err := d.cli.ContainerCreate(
+		context.Background(),
+		containerConfig,
+		hostConfig,
+		networkConfig,
+		// TODO(rabrams) namespace
+		r.Name,
+	)
+	if err != nil {
+		return volume.Response{Err: fmt.Sprintf("unexpected error: %s", err)}
+	}
+	d.cli.ContainerStart(
+		context.Background(),
+		cont.ID,
+		types.ContainerStartOptions{},
+	)
+	if err != nil {
+		return volume.Response{Err: fmt.Sprintf("unexpected error: %s", err)}
+	}
+	return volume.Response{Err: ""}
 }
 
+// List lists available volumes
 func (d ImagefsDriver) List(r volume.Request) volume.Response {
-	fmt.Printf("-> List %+v\n", r)
-	images, err := d.cli.ImageList(context.Background(), types.ImageListOptions{All: true})
+	containers, err := d.cli.ContainerList(context.Background(), types.ContainerListOptions{})
 	if err != nil {
-		panic(err)
+		return volume.Response{Err: fmt.Sprintf("unexpected error: %s", err)}
 	}
-
-	volumes := []*volume.Volume{}
-	for _, image := range images {
-		fmt.Printf("%s\n", image.ID)
-		volumes = append(volumes, &volume.Volume{Name: image.ID[7:]})
+	response := volume.Response{}
+	for i := range containers {
+		_, ok := containers[i].Labels["com.docker.imagefs.version"]
+		if !ok {
+			continue
+		}
+		response.Volumes = append(response.Volumes, &volume.Volume{
+			// TODO(rabrams) fall back to id if no names
+			Name: containers[i].Names[0],
+		})
 	}
-
-	response := volume.Response{Volumes:volumes}
-	fmt.Printf("<- %+v\n", response)
 	return response
 }
 
+// Get gets a volume
 func (d ImagefsDriver) Get(r volume.Request) volume.Response {
-	fmt.Printf("-> Get %+v\n", r)
-	inspect, _, err := d.cli.ImageInspectWithRaw(context.Background(), r.Name)
+	fmt.Printf("-> Mount %+v\n", r)
+	container, err := d.cli.ContainerInspect(context.Background(), r.Name)
 	if err != nil {
-		return volume.Response{Err: fmt.Sprint(err)}
+		return volume.Response{Err: fmt.Sprintf("unexpected error: %s", err)}
 	}
-
-	response := volume.Response{}
-	if inspect.GraphDriver.Name == "overlay2" {
-		response = volume.Response{Volume: &volume.Volume{Name: r.Name,
-			Mountpoint: strings.Split(inspect.GraphDriver.Data["UpperDir"], ":")[0]}}
-	} else {
-		response = volume.Response{Err: fmt.Sprintf("GraphDriver %s not supported", inspect.GraphDriver.Name)}
+	if container.GraphDriver.Name != "overlay" {
+		return volume.Response{Err: fmt.Sprintf("unexpected graph driver: %s", container.GraphDriver.Name)}
+	}
+	mergedDir, ok := container.GraphDriver.Data["MergedDir"]
+	if !ok {
+		return volume.Response{Err: fmt.Sprintf("missing MergedDir")}
+	}
+	// HACK directory is relative to host but docker will prepend rootfs of the
+	// plugin container
+	mergedDir = fmt.Sprintf("../../../../../../../../../../../%s", mergedDir)
```

Comment:
This won't work with the `docker service create` way of launching the plugin. Can you change it assuming that the mountpoint is relative to `/`, or make it compatible with both ways of deploying it? (Which might be possible using `PropagatedMount`, I haven't looked in detail)